### PR TITLE
cycle data and routes

### PR DESCRIPTION
**Pull request by mpolimen**
We can merge this for now, but there are some things that will be added once other code is merged from others (proper Postman checking will be done after seed data is finished as well, so it is untested):

- Strict typing using interfaces for a cycle object
- Error handling, after the error checker is merged

Additionally, I changed how exports work in mongoCollections and I set up the index file for data, so it would be helpful to get this preliminary code out for others that will be working on data code.

For decisions I made:

- Since, in the database design, we don't specify a "current cycle" field for the user; I'm going to use the convention that the last added cycle id in the user's cycles list is the current cycle. Would probably be much more efficient if we save the current cycle id in a cookie though. That's up to the person working on user stuff to implement for now
- Also, currently it is allowed to start a new cycle, even if the current cycle was not finished. To remedy this, I'm setting the current cycle's end date to today before creating a new cycle with a start date of today

### CODE REVIEW

**Comment by mpolimen** [comment]
Error handling is now added since error functions were put in main

File: server/data/cycles.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+     * @param {string} id Cycle id
+     * @returns {Promise<Object>} Cycle object
+     */
+    readById: async (id: string): Promise<object> => {
+        if(!checkObjectId(id)){
+            throw new Error('Invalid id');
+        }
+
+        const cycleCollection = await cycles();
+        const cycle = await cycleCollection.findOne({_id: new ObjectId(id)});
```
Comment (roccopolimen) [inline server/data/cycles.ts]: Add type here for variable

Comment (roccopolimen) [reply server/data/cycles.ts]: and similarly elsewhere

File: server/data/cycles.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import { checkObjectId } from '../helpers/error';
+
+export default {
+
+    /**
+     * @description Get a cycle by id
+     * @param {string} id Cycle id
+     * @returns {Promise<Object>} Cycle object
+     */
+    readById: async (id: string): Promise<object> => {
```
Comment (roccopolimen) [inline server/data/cycles.ts]: I suggest these be named more specifically, such as `getCycleById`

File: server/data/cycles.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        if(!checkObjectId(id)){
+            throw new Error('Invalid id');
+        }
+
+        const cycleCollection = await cycles();
+        const cycle = await cycleCollection.findOne({_id: new ObjectId(id)});
+        if(cycle === null)
+            throw new Error("There is no cycle with that id.");
+
+        cycle['_id'] = cycle['_id'].toString();
```
Comment (roccopolimen) [inline server/data/cycles.ts]: not sure if we will do this, since doing this will change the type of the object

File: server/data/cycles.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        let currDate: Date = new Date();
+        let cycle: object = {
+            startDate: currDate,
+            endDate: null,
+            applications: []
+        }
+
+        // Add to cycles document
+        const cycleCollection: any = await cycles();
```
Comment (roccopolimen) [inline server/data/cycles.ts]: would rather not do `:any` and just leave typeless if we do not know it. Stylistic preference, not blocking

File: server/data/index.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import cycles from './cycles';
```
Comment (roccopolimen) [inline server/data/index.ts]: I had a different vision for this based on how index files typically work, see my pull request for user data/routes for how I use it.

File: server/routes/cycles.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import express from 'express';
+import cycleData from '../data/cycles';
```
Comment (roccopolimen) [inline server/routes/cycles.ts]: Personally choosing to import by function, instead of by file.

File: server/routes/cycles.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    } catch(e) {
+        return res.status(500).json({ 
+            message: 'Internal Server Error. Failed to get current cycle.',
+            error: e.message });
+    }
+});
+
+// POST /cycles
+router.post('/', async (req, res) => {
+    // Creates a new cycle
```
Comment (roccopolimen) [inline server/routes/cycles.ts:38]: Should be checking here if express session or not before making the call

**Review by roccopolimen — COMMENTED** [review]
Quite a few comments, albeit mostly small.

File: server/data/cycles.ts (hunk 1 above)
Comment (mpolimen) [reply server/data/cycles.ts]: These would be any types, and according to your preference below, leave them type-less.

**Pull request closed without merging**